handling-forms/controlled-forms.md

```diff
@@ -0,0 +1,542 @@
+
+### New Piece of State: An Object of `formFields`
+
+Sofia needs to introduce state into the `NewStudentForm` component.
+
+She's made pieces of state that were strings and booleans before, but she's never put an object in state!
+
+To practice that, she'll make one piece of state named `formFields`. It will be an object that will hold `name` and `email` as key-value pairs.
+
+Her vision is that `formFields` will be an easy, nice way to organize the different form fields. By default, a new student's name and email should be blank. The initial state of `formFields` should be:
+
+<!-- prettier-ignore-start -->
+```js
+{
+    name: '',
+    email: ''
+}
+```
+<!-- prettier-ignore-end -->
+
+Therefore, introducing this new piece of state will look like:
+
+<!-- prettier-ignore-start -->
+```js
+import { useState } from 'react';
+
+const NewStudentForm = () => {
+
+    const [formFields, setFormFields] = useState({
+        name: '',
+        email: ''
+    });
+```
+<!-- prettier-ignore-end -->
+
+Now she needs to make the input fields read from this state. Since `formFields` is an object, we'll use dot-notation to access the `name` and `email` values.
+
+<!-- prettier-ignore-start -->
+```js
+    return (
+        <form>
+            <div>
+                <label htmlFor="fullName">Name:</label>
+                <input
+                    name="fullName"
+                    value={formFields.name}/>
+            </div>
+            <div>
+                <label htmlFor="email">Email:</label>
+                <input
+                    name="email"
+                    value={formFields.email}/>
+            </div>
+            <input
+                type="submit"
+                value="Add Student"
+            />
+        </form>
+    );
+```
+<!-- prettier-ignore-end -->
+
+### Event Handling To Update `formFields`
+
+Sofia's next step is to ensure that changes in an input field affect the component state.
+
+She can create two event handlers, `onNameChange` and `onEmailChange`. These event handlers need to do two things:
+
+- Read the current value inside the input field
+- Update state to that current value
+
+<!-- prettier-ignore-start -->
+```js
+    const onNameChange = (event) => {
+        setFormFields({
+            ...formFields,
+            name: event.target.value
+        })
+    };
+
+    const onEmailChange = (event) => {
+        setFormFields({
+            ...formFields,
+            email: event.target.value
+        })
+    };
+```
+<!-- prettier-ignore-end -->
+
+In this example, we're using [spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals) for a quick way to clone the original `formFields` object. Each event handler should add a specific key-value pair: `onNameChange` adds the key-value pair `name: event.target.value`, while `onEmailChange` adds `email`.
```

Comment:
address a possible point of confusion about why the value is `event.target.value` for both functions. and also call out the overwrite behavior the spread plus extra stuff has.
```suggestion
Sofia uses [spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals) for a quick way to clone the original `formFields` object. Each event handler should add a specific key-value pair. `onNameChange` adds the key-value pair `name: event.target.value`, while `onEmailChange` adds `email: event.target.value`.

Even though both value expressions appear the same, the event handles are registered on different inputs, so the `event.target` will refer to different input elements!

The cloned `formFields` object will already have a `name` and `email` key from the cloning, but listing the new value afterward will overwrite the cloned value.
```